addons/hw_drivers/driver.py

```diff
@@ -47,29 +52,41 @@ def supported(cls, device):
         """
         return False
 
-    def action(self, data):
+    def action(self, action='', **kwargs):
         """Helper function that calls a specific action method on the device.
 
-        :param data: the `_actions` key mapped to the action method we want to call
-        :type data: string
+        :param str action: the action to be executed
+        :param dict kwargs: the parameters to be passed to the action method
+        :return: the result of the action method
         """
-        self._actions[data.get('action', '')](data)
+        try:
+            response = {'status': 'success', 'result': self._actions[action](**kwargs), 'action_args': {**kwargs}}
+        except Exception as e:
+            _logger.exception("Error while executing action")
+            response = {'status': 'error', 'result': str(e), 'action_args': {**kwargs}}
+
+        event_manager.device_changed(self, response)  # Make response available to /event route or websocket
```

Comment:
wouldn't this conflict with the responses within the driver actions themselves?
if the driver action calls `event_manager.device_changed(self, response)` and this one too wouldn't this lead to some bugs?

Reply:
normally, all driver actions calls to `event_manager.device_changed(...)` should be removed: all events should be handled by this call.